### PR TITLE
Precompute carriage cost usage

### PR DESCRIPTION
**Pull request by michael-yxchen**
Closes: https://github.com/monad-crypto/monad-internal/issues/409

### CODE REVIEW

File: monad-eth-block-policy/src/lib.rs [hunk 1]
```diff
@@ -268,6 +274,16 @@ impl CommittedTxnBuffer {
...
+        // compute carriage cost usage in the block
+        let mut block_carriage_cost = BTreeMap::new();
+        for txn in block.validated_txns.iter() {
+            let entry = block_carriage_cost
+                .entry(EthAddress(txn.signer()))
+                .or_insert(0_u128);
+            *entry += compute_txn_carriage_cost(txn);
+        }
+        self.carriage_costs
+            .insert(block_number, block_carriage_cost);
```
Comment (mikety) [inline monad-eth-block-policy/src/lib.rs]: Assert that None is returned - i.e. no such blocks was in the cache before?

File: monad-eth-block-policy/src/lib.rs [hunk 2]
```diff
@@ -221,15 +224,14 @@ impl CommittedTxnBuffer {
...
+        for (&cache_seq_num, block_carriage_costs) in self.carriage_costs.iter() {
             if cache_seq_num > base_seq_num {
                 assert_eq!(next_validate, cache_seq_num);
-                for txn in txns {
-                    if EthAddress(txn.signer()) == *eth_address {
-                        carriage_cost += compute_txn_carriage_cost(txn);
-                    }
+                if let Some(block_carriage_cost) = block_carriage_costs.get(eth_address) {
+                    carriage_cost += block_carriage_cost;
                 }
```
Comment (mikety) [inline monad-eth-block-policy/src/lib.rs:235]: assert/warn that the expected eth_address not found in cache?
nit: its an account_carriage_cost for the committed block

Comment (michael-yxchen) [reply monad-eth-block-policy/src/lib.rs:235]: this cache stores carriage costs usage for accounts with transactions in the block. It's ok if the account we're querying doesn't have transaction in previous blocks

**Comment by michael-yxchen** [comment]
Thanks, I can add caching for pending block carriage cost in a separate PR

**Pull request merged**